guardian/pkg/asyncutil/async.go

```diff
@@ -0,0 +1,274 @@
+	go func() {
+		defer signal.Send()
+		for _, signaler := range signalers {
+			<-signaler.Receive()
+		}
+	}()
+
+	return signal
+}
+
+type commandExecutor[C any, R any] struct {
+	command             func(context.Context, C) (R, error)
+	drainAndBacklogSig  chan Signaler
+	resumeBackloggedSig Signaler
+	cmdChan             chan Command[C, R]
+	// backlogChan contains all the commands that failed with EOF, waiting to be retried.
+	backlogChan chan Command[C, R]
+	// inflightCmds keeps track of the number of commands that are currently being executed.
+	inflightCmds        sync.WaitGroup
+	executeSig          Signaler
+	shutdownCompleteSig Signaler
+	errBuff             ErrorBuffer
+
+	backLogCommands bool
+	backlog         []Command[C, R]
+}
+
+func NewCommandCoordinator(executors ...ExecutionController) ExecutionController {
+	var coordinator executorCoordinator
+	for _, executor := range executors {
+		coordinator = append(coordinator, executor)
+	}
+	return coordinator
+}
+
+// NewCommandExecutor creates a new CommandExecutor implementation. It calls the given function f with the command given
+// to Send. Any errors from the function are sent over the errBuff. If an EOF is sent over the error buff, the caller
+// must pause the executor, restart / fix whatever processes need restarting or fixing, then resume execution (using the
+// PauseExecution and ResumeExecution functions). ResumeExecution re runs the commands that failed with EOF.
+func NewCommandExecutor[C any, R any](ctx context.Context, errBuff ErrorBuffer, f func(context.Context, C) (R, error)) CommandExecutor[C, R] {
+	hdlr := &commandExecutor[C, R]{
+		command:             f,
+		errBuff:             errBuff,
+		cmdChan:             make(chan Command[C, R], 100),
+		backlogChan:         make(chan Command[C, R], 100),
+		drainAndBacklogSig:  make(chan Signaler, 100),
+		resumeBackloggedSig: NewSignaler(),
+
+		executeSig:          NewSignaler(),
+		shutdownCompleteSig: NewSignaler(),
+	}
+
+	go hdlr.loop(ctx)
+	return hdlr
+}
+
+func (executor *commandExecutor[C, R]) loop(shutdownCtx context.Context) {
+	ctx, stopCommands := context.WithCancel(shutdownCtx)
+
+	// Used to ensure we don't resume execution if we haven't finished draining. This can happen if the caller of DrainAndBacklog
+	// doesn't wait on the channel provided.
+	defer func() {
+		defer stopCommands()
+		defer executor.shutdownCompleteSig.Close()
+
+		// close the cmdChan in case anything tries to write to it. This will ensure a panic occurs while trying to
+		// clean up any outstanding cmd.
+		close(executor.cmdChan)
+
+		// Wait for inflight requests to finish before handling outstanding cmds.
+		executor.inflightCmds.Wait()
+
+		close(executor.backlogChan)
+
+		// Add all outstanding commands in the backlog or cmd channels to the backlog slice.
+		executor.backlog = append(executor.backlog, ReadAll(executor.backlogChan)...)
+		executor.backlog = append(executor.backlog, ReadAll(executor.cmdChan)...)
```

Comment:
I think we need to perform this operation before line 142: https://github.com/projectcalico/calico/pull/9849/files#diff-332e67201f01430847e13e593d0dc866cbb68781014e7dc94babff2de33d1f92R142. Either that or you do not need it.

Reply:
It doesn't need to be done before, this is actually intentional. I'm closing the channel so nothing more can be sent over it. Closing the channel doesn't "clear" it, all the items will remain on the channel until they're read off.

This basically forces a panic to happen if somebody writes to the channel after it's closed and we're trying to clean up. If they could write to it after reading everything off, then that would result in some commands hanging after close.